src/coreclr/interpreter/compiler.cpp

```diff
@@ -3325,6 +3387,28 @@ void InterpCompiler::EmitLdind(InterpType interpType, CORINFO_CLASS_HANDLE clsHn
 
 void InterpCompiler::EmitStind(InterpType interpType, CORINFO_CLASS_HANDLE clsHnd, int32_t offset, bool reverseSVarOrder)
 {
+    int stackIndexValue;
+    int stackIndexAddress;
+    if (reverseSVarOrder)
+    {
+        stackIndexAddress = -2;
+        stackIndexValue = -1;
+    }
+    else
+    {
+        stackIndexAddress = -1;
+        stackIndexValue = -2;
+    }
+
+    if (interpType == InterpTypeR4 && m_pStackPointer[stackIndexValue].type == StackTypeR8)
+    {
+        EmitConv(m_pStackPointer + stackIndexValue, StackTypeR4, INTOP_CONV_R4_R8);
+    }
+    else if (interpType == InterpTypeR8 && m_pStackPointer[stackIndexValue].type == StackTypeR4)
+    {
+        EmitConv(m_pStackPointer + stackIndexValue, StackTypeR8, INTOP_CONV_R8_R4);
+    }
```

Comment:
I think if we're in reverseSVarOrder this will de-reverse the order by putting the new converted value on top of the stack, so we will need to update stackIndexValue and stackIndexAddress. Not sure though

Reply:
Nevermind, I misunderstood EmitConv. But I think there's still a problem. It *doesn't* update the stack and instead creates a new variable and exposes it via m_pLastNewIns's dvar:
```c++
void InterpCompiler::EmitConv(StackInfo *sp, StackType type, InterpOpcode convOp)
{
    InterpInst *newInst = AddIns(convOp);

    newInst->SetSVar(sp->var);
    int32_t var = CreateVarExplicit(g_interpTypeFromStackType[type], NULL, INTERP_STACK_SLOT_SIZE);
    new (sp) StackInfo(type, NULL, var);
    newInst->SetDVar(var);
```
This new var *isn't* on the stack automatically so I don't think the rest of EmitStind is going to see it, and I think the stind will still use the old non-coerced value.

Reply:
Since stind is going to consume the stack value anyway, you could probably patch the result of the EmitConv so that dvar == svar, and then everything will work.

Reply:
the new(sp) StackInfo(type, NULL, var) line will update the stack.

Reply:
Ah right, it's overwriting the value at sp without pushing/popping. Okay. I think the reverse order thing might still behave wrong, then?

Reply:
Nevermind, I see how it works.